### PR TITLE
Change abstract stores to use the page cache

### PR DESCRIPTION
**Pull request by ragadeeshu**
Also introduces StoreVersionTrailerUtil for store version trailer handling with the page cache.
This is a step towards making the stores independent of the FilesystemAbstraction, which will help with supporting block devices.

### CODE REVIEW

File: community/kernel/src/test/java/org/neo4j/kernel/impl/storemigration/StoreMigratorTest.java [hunk 1]
```diff
@@ -74,6 +85,23 @@ public void shouldBeAbleToResumeMigration() throws Exception
...
+            public StoreChannel open( File fileName, String mode ) throws IOException
+            {
+                return new DelegatingStoreChannel( super.open( fileName, mode ) )
+                {
+                    @Override
+                    public void close() throws IOException
+                    {
+//                        new Exception( "The damn channel got closed!" ).printStackTrace();
+                        LockSupport.parkNanos( TimeUnit.MICROSECONDS.toNanos( 100 ) );
+                        super.close();
```
Comment (chrisvest) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/storemigration/StoreMigratorTest.java]: This looks weird.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java [hunk 2]
```diff
@@ -229,29 +227,43 @@ protected void loadStorage()
...
+                writeTrailer( pagedFile, new byte[expectedVersionLength], trailerOffset );
+            }
+            else
+            {
+                setStoreNotOk( new IllegalStateException(
+                        "Misaligned file size " + trailerOffset + " for " + this + ", expected version length:" +
+                        expectedVersionLength ) );
+            }
+        }
+    }
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:250]: Wondering if we can introduce a `VersionTrailer` class where to move some of this logic, e.g., length, offset, etc.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            if ( pos == sought.length )
+            {
+                return i;
+            }
+
+        }
+        return -1;
+    }
+
+    private static int getMaximumNumberOfPagesVersionSpans( int trailerLength, PagedFile pagedFile )
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java]: pass in `pageSize` instead `pagedFile`

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    {
+        return trailerLength / pagedFile.pageSize() + 2;
+    }
+
+    private static byte[] getLastBytes( PagedFile pagedFile, int maximumNumberOfPagesVersionSpans,
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java]: pass in `pageSize` instead `pagedFile`

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+            PageCursor pageCursor ) throws IOException
+    {
+        byte[] allData = new byte[pagedFile.pageSize() * maximumNumberOfPagesVersionSpans];
+        int currentPage = 0;
+        while ( pageCursor.next() )
+        {
+            byte[] data = new byte[pagedFile.pageSize()];
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java]: can we allocate this outside the while loop?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+            do
+            {
+                pageCursor.getBytes( data );
+            }
+            while ( pageCursor.shouldRetry() );
+            System.arraycopy( data, 0, allData, currentPage * data.length, data.length );
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java:171]: We can avoid this copy if we add a `getBytes( byte[] data, int writeOffset)` on the `PageCursor`

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        }
+        return version;
+    }
+
+    public static int findTrailerOffset( byte[] src, byte[] sought )
+    {
+        for ( int i = src.length - sought.length; i >= 0; i-- )
+        {
+            int pos = 0;
+            while ( pos < sought.length && src[i + pos] == sought[pos] )
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java]: before doing this, can we can skip all the trailing `0`s at the end of the `src` array?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+public abstract class StoreVersionTrailerUtil
+{
+
+    public static long getTrailerOffset( PagedFile pagedFile, String trailer )
+            throws IOException
+    {
+        int trailerOffset;
+        byte[] encodedTrailer = UTF8.encode( trailer );
+        int maximumNumberOfPagesVersionSpans = getMaximumNumberOfPagesVersionSpans( encodedTrailer.length, pagedFile );
+        long pageId = Math.max( pagedFile.getLastPageId() + 1 - maximumNumberOfPagesVersionSpans, 0 );
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java]: isn't it better having a method `computePageId` instead having `getMaximumNumberOfPagesVersionSpans` and duplicate the `long pageId = ...` line twice?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+            int writtenOffset = 0;
+            while ( writtenOffset < trailer.length )
+            {
+                pageCursor.next();
+                pageCursor.setOffset( (int) ((writtenOffset + trailerOffset) % pagedFile.pageSize()) );
+                do
+                {
+                    do
+                    {
+                        pageCursor.putByte( trailer[writtenOffset] );
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java:119]: Why not using `pageCursor.putBytes( trailer )` instead?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+            trailerOffset = findTrailerOffset( allData, encodedTrailer );
+        }
+        return trailerOffset + pageId * pagedFile.pageSize();
+    }
+
+    public static void writeTrailer( PagedFile pagedFile, byte[] trailer,
+            long trailerOffset ) throws IOException
+    {
+        long pageId = trailerOffset / pagedFile.pageSize();
+        try ( PageCursor pageCursor = pagedFile.io( pageId, PagedFile.PF_EXCLUSIVE_LOCK ) )
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java]: I am not sure I understand why `pageId` is computed this way. Shouldn't you write the trailer starting from `pagedFile.getLastPageId()`?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+package org.neo4j.kernel.impl.store;
+
+import java.io.IOException;
+
+import org.neo4j.helpers.UTF8;
+import org.neo4j.io.pagecache.PageCursor;
+import org.neo4j.io.pagecache.PagedFile;
+import org.neo4j.kernel.impl.util.Charsets;
+
+public abstract class StoreVersionTrailerUtil
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java:29]: This class needs javadocs and tests.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java (hunk 10 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java]: I figure that out, `trailerOffset` is global for the whole file not only the current page. Please add documentation this might be confusing.

Can we avoid to pass in an offset from the outside and instead expose `appendTrailer` and `replaceTrailer` and compute offsets internally?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreFile.java [hunk 12]
```diff
@@ -298,16 +297,26 @@ private static void setStoreVersionTrailer( FileSystemAbstraction fs, File targe
...
-            fileChannel.write( ByteBuffer.wrap( trailer ) );
         }
-        catch ( IllegalArgumentException e )
+        catch ( IOException e )
         {
-            throw withMessage( e, e.getMessage() + " | " + "size:" + fileSize + ", trailer:" + trailer.length +
-                    " for " + targetStoreFileName );
+            return false;
         }
+        return true;
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreFile.java:320]: I wonder if there is a better way of doing this...

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java (hunk 2 above)
Comment (ragadeeshu) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:250]: Not with offset, since that usually is computed externally, in different manners. Actually, I don't think that it is worth it, since we sometimes want to write trailers consisting of all (byte) 0.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java (hunk 10 above)
Comment (ragadeeshu) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java]: Nope, since we compute the offset in different ways depending on use-case, often with data only available externally.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java (hunk 7 above)
Comment (ragadeeshu) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java]: We do skip them, we stop the loop as soon as there is a mismatch.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java (hunk 9 above)
Comment (ragadeeshu) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java:119]: That would require either, 1. allocating new arrays and copying, or 2. a new putbytes in pagecursor with offset and length.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreFile.java (hunk 12 above)
Comment (ragadeeshu) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreFile.java:320]: Not that I know of. What we want to find out is if it is mappable by the page cache, and the way to do that is to map it.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java (hunk 8 above)
Comment (ragadeeshu) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java]: The return value from getMaximumNumberOfPagesVersionSpans is used for getting data from the right amount of pages as well.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java (hunk 2 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:250]: When?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java (hunk 7 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java]: it might have been faster to just check `0`s and the start comparing arrays, perhaps not super important

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java (hunk 8 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java]: you can compute that as `pagedFile.gerLastPageId()  - pageId + 1`

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java (hunk 9 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java:119]: I like 2.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java (hunk 10 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java]: Do we have different use cases then `append` and `replace`?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreFile.java (hunk 12 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreFile.java:320]: Would it be better to add to the page cache a method `isMappable()` to check this?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java (hunk 2 above)
Comment (ragadeeshu) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:250]: When we open/map stores (not included in this pr, but will soon appear in another pr)

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java (hunk 10 above)
Comment (ragadeeshu) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java]: Append is not a thing, since we do not want it at the end of the file, we want it after the last record, which is computed in different ways with data available locally for those classes.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreFile.java (hunk 12 above)
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreFile.java:320]: This is only temporary. The only places where we are using the value return from `storeExists`, are about whether to create the store or not. I want to change those places to pass in the wish that the store files should be mapped with the CREATE option, and then the stores can add any required header bytes the file might be missing.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java (hunk 10 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java]: I disagree on the assumption that writing a trailer should be store dependent.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreFile.java (hunk 12 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreFile.java:320]: Fair enough.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java (hunk 11 above)
Comment (ragadeeshu) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java:29]: Done

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java (hunk 6 above)
Comment (ragadeeshu) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java:171]: This will be addressed in a separate PR.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java (hunk 9 above)
Comment (ragadeeshu) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java:119]: This will be addressed in a separate PR.
